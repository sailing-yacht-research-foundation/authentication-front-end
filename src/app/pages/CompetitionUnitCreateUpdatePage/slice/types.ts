/* --- STATE --- */
export interface competitionUnitManagerState {
    lastSubscribedCompetitionUnitId: string;
}

export type ContainerState = competitionUnitManagerState;
