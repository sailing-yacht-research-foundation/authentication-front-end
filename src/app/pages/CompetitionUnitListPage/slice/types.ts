/* --- STATE --- */
export interface MyRaceListState {
    results: any[];
    page: number;
    total: number;
    is_changing_page: boolean;
  }
  
  export type ContainerState = MyRaceListState
  