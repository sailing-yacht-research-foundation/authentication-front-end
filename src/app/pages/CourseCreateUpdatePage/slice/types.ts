/* --- STATE --- */
export interface CourseState {
    course_sequenced_geometries: any[];
    page: number;
    total_course: number;
    courses: any[];
    is_handling_course: boolean;
}

export type ContainerState = CourseState;
