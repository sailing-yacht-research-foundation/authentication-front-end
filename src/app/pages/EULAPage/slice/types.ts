/* --- STATE --- */
export interface EULAState { }

export type ContainerState = EULAState
