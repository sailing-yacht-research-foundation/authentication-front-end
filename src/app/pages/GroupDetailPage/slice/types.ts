import { Group, GroupMember } from "types/Group";

/* --- STATE --- */
export interface GroupDetailState {
    adminCurrentPage: number;
    adminTotal: number;
    memberCurrentPage: number;
    memberTotal: number;
    admins: GroupMember[];
    members: GroupMember[];
    isGettingAdmins: boolean;
    isGettingMembers: boolean;
    group: Partial<Group>;
    isGettingGroup: boolean;
    getGroupDetailFailed: boolean;
    acceptedMemberResults: any[];
    adminPageSize: number;
    memberPageSize: number;
  }
  
  export type ContainerState = GroupDetailState;
  