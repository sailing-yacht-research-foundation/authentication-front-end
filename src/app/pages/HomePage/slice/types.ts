/* --- STATE --- */
export interface HomeState {
    results: any[];
    keyword: string;
    from_date: string;
    to_date: string;
    page: number;
    is_searching: boolean;
    total: number;
    page_size: number;
    show_advanced_search: boolean;
    sort: string;
  }
  
  export type ContainerState = HomeState
  