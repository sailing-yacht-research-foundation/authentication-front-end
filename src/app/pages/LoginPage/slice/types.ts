/* --- STATE --- */
export interface LoginState {
  user: any,
  is_authenticated: boolean,
  session_token: string,
  syrf_authenticated: boolean;
  user_coordinate?: null | Coordinate;
}

export type ContainerState = LoginState

export interface Coordinate {
  lat: number;
  lon: number;
}