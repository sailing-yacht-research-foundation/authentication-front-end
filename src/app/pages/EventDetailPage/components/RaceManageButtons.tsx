import React from 'react';
import { Space, Tooltip } from 'antd';
import { BorderedButton, CreateButton } from 'app/components/SyrfGeneral';
import { EventState, RaceStatus } from 'utils/constants';
import { FiEdit } from 'react-icons/fi';
import { useHistory } from 'react-router-dom';
import { translations } from 'locales/translations';
import { useTranslation } from 'react-i18next';
import { RegisterRaceModal } from 'app/components/RegisterRaceModal';
import { CompetitionUnit } from 'types/CompetitionUnit';
import { CalendarEvent } from 'types/CalendarEvent';
import { StopRaceConfirmModal } from 'app/pages/MyEventCreateUpdatePage/components/modals/StopRaceConfirmModal';

interface IRaceManageButtons {
    race: CompetitionUnit,
    canManageEvent: Function,
    event: CalendarEvent,
    setCompetitionUnit: Function,
    showRegisterModal: boolean,
    setShowRegisterModal: Function,
    isAuthenticated: boolean,
    showDeleteRaceModal: Function,
    relations: any[],
    reloadParent?: Function
}

export const RaceManageButtons = (props: IRaceManageButtons) => {
    const { race, reloadParent, canManageEvent, event, setCompetitionUnit,
        showRegisterModal, setShowRegisterModal, isAuthenticated, showDeleteRaceModal, relations } = props;

    const history = useHistory();

    const [relation, setRelation] = React.useState<any>(null);

    const [showStopRaceConfirmModal, setShowStopRaceConfirmModal] = React.useState<boolean>(false);

    const { t } = useTranslation();

    const showRegisterModalOrRedirect = (competitionUnit: CompetitionUnit) => {
        if (isAuthenticated) {
            setCompetitionUnit(competitionUnit);
            setShowRegisterModal(true);
        } else {
            history.push('/signin');
        }
    }

    const openStopRaceConfirmModal = (competitionUnit: CompetitionUnit) => {
        setCompetitionUnit(competitionUnit);
        setShowStopRaceConfirmModal(true);
    }

    const canDeleteRace = () => {
        return [RaceStatus.SCHEDULED].includes(race.status);
    }

    React.useEffect(() => {
        if (relations.length > 0) {
            setRelation(relations.find(r => r.id === race.id));
        }
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [relations]);

    const canRegisterToRace = () => { // race is scheduled and event is open and allow for registration and event status is on going or scheduled and the user is not admin.
        const isNotAdminOrParticipant = relation && !relation.isAdmin && !relation.isParticipating;
        const eventIsRegattaAndOngoingOrScheduled = event.isOpen && event.allowRegistration && [EventState.ON_GOING, EventState.SCHEDULED].includes(event.status!);
        const isNotEventEditor = !event.isEditor;
        const raceIsScheduled = [RaceStatus.SCHEDULED].includes(race.status!);

        return isNotAdminOrParticipant && eventIsRegattaAndOngoingOrScheduled && isNotEventEditor && raceIsScheduled;
    }

    const canStopRace = () => {
        const isAdmin = relation?.isAdmin;
        const isRaceOnGoing = [RaceStatus.ON_GOING].includes(race.status!);

        return isAdmin && isRaceOnGoing;
    }

    return (<Space size="middle">
        <StopRaceConfirmModal reloadParent={reloadParent} race={race} showModal={showStopRaceConfirmModal} setShowModal={setShowStopRaceConfirmModal} />
        <RegisterRaceModal
            showModal={showRegisterModal}
            setShowModal={setShowRegisterModal}
            raceName={race.name}
            lon={event.lon}
            lat={event.lat}
            setRelation={setRelation}
            raceId={race.id}
            eventId={race.calendarEventId}
        />
        {canRegisterToRace() && <CreateButton icon={<FiEdit style={{ marginRight: '10px' }} />} onClick={() => showRegisterModalOrRedirect(race)}>{t(translations.home_page.register_as_competitor)}</CreateButton>}
        {canStopRace() && <CreateButton onClick={() => openStopRaceConfirmModal(race)}>{t(translations.competition_unit_list_page.stop)}</CreateButton>}
        {canManageEvent() && <>
            <Tooltip title={t(translations.tip.update_race)}>
                <BorderedButton onClick={() => {
                    history.push(`/events/${race.calendarEventId}/races/${race.id}/update`);
                }} type="primary">{t(translations.general.update)}</BorderedButton>
            </Tooltip>

            { canDeleteRace() && <Tooltip title={t(translations.tip.delete_race)}>
                <BorderedButton danger onClick={() => showDeleteRaceModal(race)}>{t(translations.general.delete)}</BorderedButton>
            </Tooltip>}
        </>}
    </Space>)
}