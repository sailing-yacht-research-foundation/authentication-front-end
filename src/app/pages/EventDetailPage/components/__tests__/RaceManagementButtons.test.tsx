import { render } from '@testing-library/react';
import MyProvider from 'app/components/Provider';
import { i18n } from 'locales/i18n';
import { translations } from 'locales/translations';
import React from 'react';
import { createRenderer } from 'react-test-renderer/shallow';
import { CalendarEvent } from 'types/CalendarEvent';
import { CompetitionUnit } from 'types/CompetitionUnit';
import { EventState, RaceStatus } from 'utils/constants';
import { RaceManageButtons } from '../RaceManageButtons';

const shallowRenderer = createRenderer();

describe('RaceManagementButtons', () => {
    it('should render and match snapshot', () => {
        const race: CompetitionUnit = {} as CompetitionUnit;
        const event: CalendarEvent = {} as CalendarEvent;
        shallowRenderer.render(
            <MyProvider>
                <RaceManageButtons race={race} reloadParent={jest.fn} event={event} setCompetitionUnit={jest.fn} showDeleteRaceModal={jest.fn} />
            </MyProvider>);
        const renderedOutput = shallowRenderer.getRenderOutput();
        expect(renderedOutput).toMatchSnapshot();
    });

    it('should render stop race button if the condition matches', async () => {
        const race: CompetitionUnit = { status: RaceStatus.ON_GOING } as CompetitionUnit;
        const event: CalendarEvent = { isEditor: true } as CalendarEvent;
        const t = await i18n;

        const { getByText } = render(<MyProvider>
            <RaceManageButtons race={race} reloadParent={jest.fn} event={event} setCompetitionUnit={jest.fn} showDeleteRaceModal={jest.fn} />
        </MyProvider>);

        expect(getByText(String(t(translations.general.stop)))).toBeInTheDocument();
    });

    it('should render update button if the condition matches', async () => {
        const race: CompetitionUnit = {} as CompetitionUnit;
        const event: CalendarEvent = { isEditor: true, status: EventState.DRAFT } as CalendarEvent;
        const t = await i18n;
        const { getByRole } = render(<MyProvider>
            <RaceManageButtons race={race} reloadParent={jest.fn} event={event} setCompetitionUnit={jest.fn} showDeleteRaceModal={jest.fn} />
        </MyProvider>);

        expect(getByRole('button', t(translations.tip.update_race))).toBeInTheDocument();
    });

    it('should render delete button if the condition matches', async () => {
        const race: CompetitionUnit = { status: RaceStatus.SCHEDULED } as CompetitionUnit;
        const event: CalendarEvent = { isEditor: true, status: EventState.DRAFT } as CalendarEvent;
        const { container } = render(<MyProvider>
            <RaceManageButtons race={race} reloadParent={jest.fn} event={event} setCompetitionUnit={jest.fn} showDeleteRaceModal={jest.fn} />
        </MyProvider>);
    
        expect(container.querySelector('.ant-btn-dangerous')).toBeInTheDocument();
    });
});