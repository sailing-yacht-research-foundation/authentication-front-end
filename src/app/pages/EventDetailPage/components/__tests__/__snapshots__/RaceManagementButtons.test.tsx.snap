// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RaceManagementButtons should render and match snapshot 1`] = `
<Provider
  store={
    Object {
      "createReducer": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "injectedReducers": Object {},
      "injectedSagas": Object {},
      "replaceReducer": [Function],
      "runSaga": [Function],
      "subscribe": [Function],
      Symbol(Symbol.observable): [Function],
    }
  }
>
  <RaceManageButtons
    event={Object {}}
    race={Object {}}
    reloadParent={[Function]}
    setCompetitionUnit={[Function]}
    showDeleteRaceModal={[Function]}
  />
</Provider>
`;
